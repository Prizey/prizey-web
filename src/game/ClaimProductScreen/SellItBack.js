import React, { Fragment } from 'react'
import { withStyles } from '@material-ui/core/styles'
import { Button, Typography } from '@material-ui/core'
import CircularProgress from '@material-ui/core/CircularProgress'

import PlayAgain from 'design/PlayAgain'
import TransactionComponent from 'game/TransactionComponent'

const TICKET_AMOUNT = 3

const styles = theme => ({
  icon: {
    height: theme.spacing.md,
  },
  quantity: {
    marginLeft: theme.spacing.xs,
    marginRight: theme.spacing.xs,
    width: theme.spacing.sm,
  },
  rightIcon: {
    alignItems: 'center',
    display: 'flex',
    height: theme.spacing.md,
    position: 'absolute',
    right: 0,
  },
})

export const handleConfirm = ({ creating, create, amount }) => () => {
  if (!creating) {
    create({ amount })
  }
}

export class SellItBackComponent extends React.Component {
  state = {
    dialogIsOpen: false,
  }

  renderError = ({ error }) =>
    error && (
      <Typography color="error" align="center" variant="h6">
        {error}
      </Typography>
    )

  renderButton = ({ creating, classes }) => (
    <Button
      aria-label="Sell it Back"
      variant="contained"
      color="secondary"
      fullWidth
      onClick={() => this.setState({ dialogIsOpen: true })}
    >
      {creating ? (
        <CircularProgress color="inherit" size={30} />
      ) : (
        <Fragment>
          <span>SELL IT BACK</span>
          <div className={classes.rightIcon}>
            <img
              src="/icons/diamond.png"
              alt="diamond"
              className={classes.icon}
            />
            <span className={classes.quantity}>{TICKET_AMOUNT}</span>
          </div>
        </Fragment>
      )}
    </Button>
  )

  render() {
    const { dialogIsOpen } = this.state

    return (
      <Fragment>
        {this.renderError(this.props)}
        <PlayAgain
          aria-label="Play again modal"
          isOpen={dialogIsOpen}
          close={() => this.setState({ dialogIsOpen: false })}
          confirm={handleConfirm(this.props)}
        >
          You&apos;ll sell the product and get diamonds. This can&apos;t be
          undone.
        </PlayAgain>

        {this.renderButton(this.props)}
      </Fragment>
    )
  }
}

export const afterCreate = ({
  navigate,
  currentUser,
  setCurrentUser,
}) => () => {
  setCurrentUser({
    ...currentUser,
    tickets: currentUser.tickets + TICKET_AMOUNT,
  })
  navigate('/sold-back')
}

export default withStyles(styles)(props => (
  <TransactionComponent
    render={renderProps => <SellItBackComponent {...props} {...renderProps} />}
    afterCreate={afterCreate(props)}
  />
))
