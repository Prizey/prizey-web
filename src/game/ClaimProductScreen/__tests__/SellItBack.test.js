import React from 'react'
import renderer from 'react-test-renderer'
import SellItBack, { afterCreate, createInnerComponent } from '../SellItBack'

jest.mock('croods', () => ({
  New: props => (
    <div {...props}>
      New - <div>render - {props.render(props)}</div>
    </div>
  ),
}))

it('renders correctly', () => {
  const tree = renderer
    .create(<SellItBack amount={3} clearProduct={jest.fn} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('call the clear on after create', () => {
  const clearProduct = jest.fn()
  afterCreate({ clearProduct })()

  expect(clearProduct).toHaveBeenCalledTimes(1)
})

it('call the create when click on button', () => {
  const Component = createInnerComponent({ amount: 3, classes: {} })
  const params = {
    create: jest.fn(),
    creating: false,
    error: null,
  }
  const tree = renderer.create(<Component {...params} />).root
  tree.findByProps({ 'aria-label': 'action-button' }).props.onClick()

  expect(params.create).toHaveBeenCalledTimes(1)
})
