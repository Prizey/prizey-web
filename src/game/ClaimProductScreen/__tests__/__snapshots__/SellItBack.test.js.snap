// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  afterCreate={[Function]}
  name="tickets"
  path="/ticket_transactions"
  render={[Function]}
>
  New - 
  <div>
    render - 
    <button
      aria-label="action-button"
      className="MuiButtonBase-root-30 MuiButton-root-4 MuiButton-contained-15 MuiButton-containedSecondary-17 MuiButton-raised-18 MuiButton-raisedSecondary-20 MuiButton-fullWidth-29"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-5"
      >
        <span>
          SELL IT BACK
        </span>
        <div
          className="Component-rightIcon-3"
        >
          <img
            alt="diamond"
            className="Component-icon-1"
            src="/icons/diamond.png"
          />
          <span
            className="Component-quantity-2"
          >
            3
          </span>
        </div>
      </span>
      <span
        className="MuiTouchRipple-root-33"
      />
    </button>
  </div>
</div>
`;
